numpy/lib/function_base.py

```diff
@@ -1569,22 +1608,31 @@ def gradient(f, *varargs, **kwargs):
     if max(axes) >= N or min(axes) < 0:
         raise ValueError("'axis' entry is out of bounds")
 
-    if len(set(axes)) != len(axes):
+    len_axes = len(axes)
+    if len(set(axes)) != len_axes:
         raise ValueError("duplicate value in 'axis'")
 
     n = len(varargs)
     if n == 0:
-        dx = [1.0]*N
-    elif n == 1:
-        dx = [varargs[0]]*N
-    elif n == len(axes):
+        dx = [1.0] * len_axes
+    elif n == len_axes or (n == 1 and np.isscalar(varargs[0])):
         dx = list(varargs)
+        for i, distances in enumerate(dx):
```

Comment:
I think this is a step in the right direction. I'll have more detailed suggestions, but let's settle on the overall API first.